Source/Core/DolphinQt/Settings/GameCubePane.cpp

```diff
@@ -303,66 +303,89 @@ void GameCubePane::BrowseMemcard(ExpansionInterface::Slot slot)
+    if (File::Exists(path))
     {
-      QString path_other =
-          QFileInfo(QString::fromStdString(Config::Get(Config::GetInfoForMemcardPath(other_slot))))
-              .absoluteFilePath();
+      auto [error_code, mc] = Memcard::GCMemcard::Open(path);
 
-      if (path_abs == path_other)
+      if (error_code.HasCriticalErrors() || !mc || !mc->IsValid())
       {
         ModalMessageBox::critical(
             this, tr("Error"),
-            tr("The same file can't be used in multiple slots; it is already used by %1.")
-                .arg(QString::fromStdString(fmt::to_string(other_slot))));
+            tr("The file\n%1\nis either corrupted or not a GameCube memory card file.\n%2")
+                .arg(QString::fromStdString(path))
+                .arg(GCMemcardManager::GetErrorMessagesForErrorCode(error_code)));
         return;
       }
     }
   }
 
-  QString path_old =
-      QFileInfo(QString::fromStdString(Config::Get(Config::GetInfoForMemcardPath(slot))))
-          .absoluteFilePath();
+  // Check if the other slot has the same memory card configured and refuse to use this card if so.
+  // The EU path is compared here, but it doesn't actually matter which one we compare since they
+  // follow a known pattern, so if the EU path matches the other match too and vice-versa.
+  for (ExpansionInterface::Slot other_slot : ExpansionInterface::MEMCARD_SLOTS)
+  {
+    if (other_slot == slot)
+      continue;
+
+    const std::string other_eu_path = Config::GetMemcardPath(other_slot, DiscIO::Region::PAL);
+    if (eu_path == other_eu_path)
+    {
+      ModalMessageBox::critical(
+          this, tr("Error"),
+          tr("The same file can't be used in multiple slots; it is already used by %1.")
+              .arg(QString::fromStdString(fmt::to_string(other_slot))));
+      return;
+    }
+  }
 
-  Config::SetBase(Config::GetInfoForMemcardPath(slot), path_abs.toStdString());
+  Config::SetBase(Config::GetInfoForMemcardPath(slot), raw_path);
 
-  if (Core::IsRunning() && path_abs != path_old)
+  if (Core::IsRunning())
   {
-    // ChangeDevice unplugs the device for 1 second, which means that games should notice that
-    // the path has changed and thus the memory card contents have changed
-    ExpansionInterface::ChangeDevice(slot, ExpansionInterface::EXIDeviceType::MemoryCard);
+    // If emulation is running and the new card is different from the old one, notify the system to
+    // eject the old and insert the new card.
+    // TODO: This should probably done by a config change callback instead.
+    const std::string old_eu_path = Config::GetMemcardPath(slot, DiscIO::Region::PAL);
+    if (eu_path != old_eu_path)
+    {
+      // ChangeDevice unplugs the device for 1 second, which means that games should notice that
+      // the path has changed and thus the memory card contents have changed
+      ExpansionInterface::ChangeDevice(slot, ExpansionInterface::EXIDeviceType::MemoryCard);
+    }
```

Comment:
This whole thing seems... unfinished? Why is it only checking PAL? Explains swapping being broken, as I'm using USA memcards.

Reply:
> Why is it only checking PAL?

FWIW, I asked that question above and they put a comment, it's likely the same reason:

> // The EU path is compared here, but it doesn't actually matter which one we compare since they
  // follow a known pattern, so if the EU path matches the other match too and vice-versa.

Reply:
Ah, I see. It was in a comment a bit above this block of code so I missed it. Seems I was right on the second point though.